## Group SQL Practice

_Author: Matt Brems & Katie Sylvia_

### Required Preparation

In order to run SQL queries within Python/this Jupyter notebook, you will need to install the following:

```bash
conda install sqlalchemy
conda install psycopg2
```

If you get an error trying to install `psycopg2`, you can install a stand-alone package (sufficient for everything you'll need to complete this lab) by entering the following anywhere in the terminal: 

```bash 
pip install psycopg2-binary
```

(If `pip` doesn't work for you, try `pip3 install psycopg2-binary`.)

In [1]:
! pip install psycopg2-binary


[notice] A new release of pip is available: 23.1.2 -> 23.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


### Instructions

In this lab, you're going to have ten prompts. Each prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

The purpose of this lab is to get you to practice your SQL - an **important** skill in data science! While it is possible to get these answers using Pandas, it should be your goal to do everything only in SQL.

**Data**: The data used in this lab is the Iowa liquor database. It contains various tables of transactions. The schema for this database is here:

<img src="./images/schema.png" alt="schema" width="750"/>

In [27]:
# Import Pandas and Create_Engine

import pandas as pd
from sqlalchemy import create_engine

# Set up the engine to access the data.
engine = create_engine('postgresql://analytics_student:analyticsga@analyticsga-psql.generalassemb.ly:5432/iowa_liquor_sales_database')

### Prompt 1

How many rows are there in the `stores` table of the Iowa liquor database?

In [28]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT COUNT(*) AS num_rows
FROM stores
"""

pd.read_sql_query(sql, engine)

num_rows
0      1973

**Prompt 1 Answer**:

There are 1,973 rows.

### Prompt 2

If you sort the stores in alphabetical order by name, which store is fourth?

In [29]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT name
FROM stores
ORDER BY name
LIMIT 4
"""

pd.read_sql_query(sql, engine)

name
0  218 Fuel Express & Chubby's Liquor
1          3047 Cub Foods / Iowa City
2               3060 Cub Foods / Ames
3         3061 Cub Foods / Sioux City

**Prompt 2 Answer**:

Store 3061 Cub Foods/ Sioux City is fourth.

### Prompt 3

Among those sales whose category name is `IMPORTED VODKA`, if you sort by vendor in alphabetical order, what is the bottle price and number of bottles bought in the first 3 transactions?
> You should give us three sets of numbers - one for each transaction.

In [30]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT btl_price, bottle_qty, vendor
FROM sales
WHERE category_name = 'IMPORTED VODKA' 
ORDER BY vendor LIMIT 3
"""

pd.read_sql_query(sql, engine)

btl_price  bottle_qty                     vendor
0     $8.99          12  Adamba Imports Int'l Inc.
1    $28.50           6      A Hardy / U.S.A. Ltd.
2    $28.50           3      A Hardy / U.S.A. Ltd.

**Prompt 3 Answer**: 


### Prompt 4

What is the total amount of liquor sold in February 2015?

In [6]:

sql = """
SELECT SUM(total) AS total_amt_sold
FROM sales
WHERE date >= '2015-02-01' 
AND date <='2015-03-01';

"""
pd.read_sql_query(sql, engine).map('{:,.0f}'.format)

total_amt_sold
0     21,295,351

**Prompt 4 Answer**:

The total amount of liquor sold in February 2015 was 21,295,351.

### Prompt 5

Among all transactions where `IMPORTED VODKA` was sold, which vendor has the most transactions? Report the total number of transactions, the number of bottles sold, and the total amount of revenue from these transactions. Rename the columns to make sure there isn't any confusion as to what they mean!

In [31]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT vendor,
COUNT(*) AS transaction_count,
SUM(bottle_qty) AS total_bottle_sold,
SUM(total) AS revenue_of_transaction
FROM sales
WHERE category_name = 'IMPORTED VODKA'
GROUP BY vendor
ORDER BY transaction_count DESC
LIMIT 1
"""

pd.read_sql_query(sql, engine)

vendor  transaction_count  total_bottle_sold  \
0  Pernod Ricard USA/Austin Nichols              42338             403450   

   revenue_of_transaction  
0              7880365.88

**Prompt 5 Answer**: 

Pernod Ricard USA/Austin Nichols had the most transactions.
The total number of transcations were 42338.
The total number of bottles wold were 403450.
The total amount of revenue from these transactions were 7880365.88.

### Prompt 6

The reason we sell things is to make money. In sales, the term **markup** means the amount of extra money charged for a product over the cost to make that product. (You can think of the markup as the profit for that product.)

In Iowa, the law states the minimum price at which liquor may be sold.

Calculate the markup (name this `Markup`) that shows the difference between the `btl_price` and the `state_btl_cost`. 

> If you haven't subtracted columns together before, [this StackOverflow question](https://stackoverflow.com/questions/7536996/multiplying-two-columns-in-sql-server/7537059) may be helpful.

In [8]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT btl_price - state_btl_cost AS markup
FROM sales
"""

pd.read_sql_query(sql, engine)

markup
0         $5.75
1         $5.75
2         $5.75
3         $8.58
4         $2.50
...         ...
3049908   $5.75
3049909  $10.00
3049910   $5.75
3049911   $5.75
3049912  $10.00

[3049913 rows x 1 columns]

### Prompt 7

Calculate the "Percentage Markup" by dividing markup by the state bottle cost. (Rather than the total profit per bottle, this will tell you how much stores are increasing their price over the state minimum.)

After calculating the "Percentage Markup," calculate the average percentage markup per **store**. Which store has the largest average markup, and what is that average markup?

In [17]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT st.name,
AVG((sa.btl_price - sa.state_btl_cost) / sa.state_btl_cost) AS avg_percent_markup
FROM stores st, sales sa
WHERE st.store = sa.store
GROUP BY st.name
ORDER BY avg_percent_markup DESC
"""

pd.read_sql_query(sql, engine)

name  avg_percent_markup
0                   Wal-Mart 0841 / Tipton            0.525471
1               Wal-Mart 1546 / Iowa Falls            0.521499
2               Wal-Mart 1683 / Shenandoah            0.518903
3                 Wal-Mart 3394 / Atlantic            0.515094
4                Wal-Mart 2935 / Knoxville            0.514591
...                                    ...                 ...
1300                      Hwy 34 Truckstop            0.500011
1301           Paradise Distilling Company            0.500000
1302  Hawkeye Convenience Store Wiley Blvd            0.499994
1303                           D & B Foods            0.499781
1304  Hawkeye Convenience Stores / 16th St            0.499214

[1305 rows x 2 columns]

**Prompt 7 Answer**: 

Wal-Mart 0841 / Tipton has the largest average markeup with an average markup of 0.525471.

### Prompt 8

Building off of your last query, retrieve the five stores with the highest average markup percentage. Along with the store IDs and average markup percentage, please return the minimum and maximum markup percentage for those stores, as well as the names and addresses of those store.

What are the names of the five stores?

In [20]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT st.name, st.store_address, st.store,
AVG((sa.btl_price - sa.state_btl_cost) / sa.state_btl_cost) AS avg_percent_markup,
MIN((sa.btl_price - sa.state_btl_cost) / sa.state_btl_cost) AS min_percent_markup,
MAX((sa.btl_price - sa.state_btl_cost) / sa.state_btl_cost) AS max_percent_markup
FROM stores st, sales sa
WHERE st.store = sa.store
GROUP BY st.name, st.store_address, st.store
ORDER BY avg_percent_markup DESC
LIMIT 5
"""

pd.read_sql_query(sql, engine)

name  \
0      Wal-Mart 0841 / Tipton   
1  Wal-Mart 1546 / Iowa Falls   
2  Wal-Mart 1683 / Shenandoah   
3    Wal-Mart 3394 / Atlantic   
4   Wal-Mart 2935 / Knoxville   

                                       store_address  store  \
0  1126 Highway 38 North\nTipton, IA 527720000\n(...   4013   
1  840 S Oak\nIowa Falls, IA 501260000\n(42.50295...   4024   
2  705 S Fremont\nShenandoah, IA 516010000\n(40.7...   4266   
3  1905 East 7th St\nAtlantic, IA 500220000\n(41....   3833   
4  814 W Bell Ave\nKnoxville, IA 501380000\n(41.3...   3660   

   avg_percent_markup  min_percent_markup  max_percent_markup  
0            0.525471            0.498571            6.708763  
1            0.521499            0.498462            6.708763  
2            0.518903            0.498462            6.708763  
3            0.515094            0.498462            6.708763  
4            0.514591            0.498462            6.708763

**Prompt 8 Answer**:

The names of the five stores with the highest average markup percentage are:
Wal-Mart 0841 / Tipton
Wal-Mart 1546 / Iowa Falls
Wal-Mart 1683 / Shenandoah
Wal-Mart 3394 / Atlantic	
Wal-Mart 2935 / Knoxville	

### Prompt 9

Which nine counties sell the highest number of bottles of liquor per capita (per person)? Calculate the average bottle size of the bottles of liquor sold by these counties rounded to the nearest first decimal place. Of the nine counties selling the highest number of bottles of liquor per capita, which county sells (on average) the largest bottle size?

**_BONUS FUN_**: This question can (also) be solved with a SQL subquery - that is, a query that queries a query! If you'd like to give this a try, you can read more about it [here](https://www.dofactory.com/sql/subquery). Otherwise, keep an eye out for the solution key!

In [33]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT c.county,
(SUM(s.bottle_qty))/c.population AS bottles_per_capita,
round(AVG(p.bottle_size),1) AS avg_bottle_size
FROM counties c, sales s, products p
WHERE s.county = c.county
AND p.item_no = s.item
GROUP BY c.county
ORDER BY bottles_per_capita
DESC
LIMIT 10;
"""

pd.read_sql_query(sql, engine)

county  bottles_per_capita  avg_bottle_size
0      Dickinson                  22            974.1
1           Polk                  15            892.6
2     Black Hawk                  15            862.9
3    Cerro Gordo                  14            986.4
4          Scott                  13            871.6
5        Johnson                  13            895.4
6           Linn                  12            886.3
7        Kossuth                  11           1041.0
8  Pottawattamie                  11            893.7
9        Carroll                  10           1005.7

**Prompt 9 Answer**:

Dickinson sold the most per capita.
Kossuth sold the largest size of bottle of the nine counties.

### Prompt 10 (advanced but required!)

In the `sales` table, three columns are `btl_price`, `bottle_qty`, and `total`. The `total` for a transaction _should be_ the product of `btl_price` and `bottle_qty`. How many transactions have a value of `total` that is not equal to `btl_price` time `bottle_qty`?

> There will be a type error as well! We cannot compare type `money` to type `real`. We have to convert `total` to type `money` so that we can compare the two directly. [This link](https://docs.microsoft.com/en-us/sql/t-sql/functions/cast-and-convert-transact-sql?view=sql-server-ver15#syntax) may be helpful to you!

In [26]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT count(vendor)
FROM sales
WHERE total::numeric <> (btl_price*bottle_qty)::numeric
LIMIT 10
"""

pd.read_sql_query(sql, engine)

count
0      0

**Prompt 10 Answer**:

0 transactions have a value of total that is not equal to btl_price time bottle_qty.